### PR TITLE
Fix: 'NoneType' object has no attribute 'generate_signed_url' (#3226)

### PR DESCRIPTION
**Pull request by jessebrennan**
Attached to issue: https://github.com/DataBiosphere/azul/issues/3226

Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [ ] ~`make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>~
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in sandbox and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by jessebrennan** [comment]
Some proof that this works with @hannes-ucsc's reproduction

```
(.venv) jesse@vader ~/gi/azul (issues/jesse/3226-drs)$ git log --oneline 
a2b78309 (HEAD -> issues/jesse/3226-drs, github/issues/jesse/3226-drs) Return correct AccessMethod if access_id is used
006a47a7 Fix: 'NoneType' object has no attribute 'generate_signed_url' (#3226)
6e255c3b (github/develop, github/HEAD) Fix: Stitching trips over mixed links schema versions (#3203, PR #3216)
(.venv) jesse@vader ~/gi/azul (issues/jesse/3226-drs)$ python - << EOF
from azul.modules import load_app_module
s = load_app_module('service')
a = s.app
c = a.repository_controller
c.download_file(catalog='dcp2', fetch=True, file_uuid='0cff5b04-afcf-5215-bdcb-d11d994e6459',query_params={'version':'2021-02-10T16:56:40.419579Z'}, headers={})
EOF
(.venv) jesse@vader ~/gi/azul (issues/jesse/3226-drs)$ echo $?
0
```

File: src/azul/drs.py [hunk 1]
```diff
@@ -64,8 +64,8 @@ def __str__(self) -> str:
 
 @attr.s(auto_attribs=True, kw_only=True, frozen=True)
 class Access:
-    method: AccessMethod
     url: str
+    method: AccessMethod = AccessMethod.https
```
Comment (hannes-ucsc) [inline src/azul/drs.py:68]: I don't think the default applies generally. It's entirely possible that `GET /objects/{object_id}/access/{access_id}` returns a `file:` URL.

File: src/azul/drs.py [hunk 2]
```diff
@@ -147,8 +147,9 @@ def _get_object_access(self,
             response = self._request(url)
             if response.status == 200:
                 response = json.loads(response.data)
-                return Access(method=access_method,
-                              url=response['url'],
+                # GET /objects/{object_id}/access/{access_id} always gives a
```
Comment (hannes-ucsc) [inline src/azul/drs.py:150]: At the very best this is specific to TDRs implementation of DRS.

File: src/azul/drs.py [hunk 3]
```diff
@@ -123,13 +123,13 @@ def _get_object(self, drs_uri: str, access_method: AccessMethod) -> Access:
...
-                if access_id is None:
-                    access_url = method['access_url']
-                    require(access_url is not None)
+                access_url = method.get('access_url')
+                if access_url:
                     return Access(method=access_method,
                                   url=access_url['url'])
                 else:
+                    access_id = method['access_id']
+                    require(access_id is not None)
```
Comment (hannes-ucsc) [inline src/azul/drs.py:132]: This doesn't accurately address https://github.com/databiosphere/azul/issues/3226#issuecomment-878516143

> Tighten requirements around parsing the DRS URL

The main issue is that the caller (in the reproduction) continues to parse the URL without even checking that it is a `gs://` URL.

File: src/azul/drs.py [hunk 4]
```diff
@@ -147,8 +147,9 @@ def _get_object_access(self,
             response = self._request(url)
             if response.status == 200:
                 response = json.loads(response.data)
-                return Access(method=access_method,
```
Comment (hannes-ucsc) [inline src/azul/drs.py:150]: Even if we can assume that a URL returned by GET /objects/{object_id}/access/{access_id} is always `https`—which I don't think we can, the standard says nothing about that—we should encode that here, not as a default in the Access class.

**Review by hannes-ucsc — COMMENTED** [review]
To speed this up, I will just make the changes.

**Comment by hannes-ucsc** [comment]
I spot-checked this against both `dev` and `prod` but didn't run the IT, and neither did @jessebrennan it seems. We'll learn during the sandbox build if this actually fixes the IT.

**Pull request merged**